frontend/catalyst/pennylane_extensions.py

```diff
@@ -2633,22 +2635,38 @@ def callback_implementation(
     flat_args, in_tree = tree_flatten((args, kwargs))
     metadata = CallbackClosure(args, kwargs)
 
+    def to_shaped_array(ty):
+        if ty == inspect.Signature.empty:
+            return None
+        if isinstance(ty, ShapedArray):
+            return ty
+        return shaped_abstractify(ty)
+
+    results_aval = tree_map(to_shaped_array, result_shape_dtypes)
+    flat_results_aval, out_tree = tree_flatten(results_aval)
+
     def _flat_callback(flat_args):
         """Each flat_arg is a pointer.
 
         It is a pointer to a memref object.
         To find out which element type it has, we use the signature obtained previously.
         """
         jnpargs = metadata.getArgsAsJAXArrays(flat_args)
-
         args, kwargs = tree_unflatten(in_tree, jnpargs)
-        return tree_leaves(cb(*args, **kwargs))
+        retvals = tree_leaves(cb(*args, **kwargs))
```

Comment:
Is it the same as `leaves, _ = tree_flatten(...)`?

Reply:
[Yes](https://github.com/google/jax/blob/a4cccd98c9aab53746d0155ce8db2df9ef192bdf/jax/_src/tree_util.py#L160)

```python
  return default_registry.flatten(tree, is_leaf)[0]
```